src/hotspot/cpu/aarch64/aarch64_vector_ad.m4

```diff
@@ -0,0 +1,4693 @@
+  format %{ "reduce_add$1_masked $dst_src1, $pg, $dst_src1, $src2" %}
+  ins_encode %{
+    __ sve_fadda($dst_src1$$FloatRegister, __ $2,
+                 $pg$$PRegister, $src2$$FloatRegister);
+  %}
+  ins_pipe(pipe_slow);
+%}')dnl
+dnl
+REDUCE_ADD_INT_PREDICATE(I, iRegIorL2I)
+REDUCE_ADD_INT_PREDICATE(L, iRegL)
+REDUCE_ADD_FP_PREDICATE(F, S)
+REDUCE_ADD_FP_PREDICATE(D, D)
+
+// ------------------------------ Vector reduction mul -------------------------
+
+instruct reduce_mulI(iRegINoSp dst, iRegIorL2I isrc, vReg vsrc,
+                     vReg tmp1, vReg tmp2) %{
+  predicate(Matcher::vector_length_in_bytes(n->in(2)) == 8 ||
+            Matcher::vector_length_in_bytes(n->in(2)) == 16);
+  match(Set dst (MulReductionVI isrc vsrc));
+  effect(TEMP_DEF dst, TEMP tmp1, TEMP tmp2);
+  format %{ "reduce_mulI $dst, $isrc, $vsrc\t# vector (64/128 bits). KILL $tmp1, $tmp2" %}
+  ins_encode %{
+    BasicType bt = Matcher::vector_element_basic_type(this, $vsrc);
+    uint length_in_bytes = Matcher::vector_length_in_bytes(this, $vsrc);
+    __ neon_reduce_mul_integral($dst$$Register, bt, $isrc$$Register,
+                                $vsrc$$FloatRegister, length_in_bytes,
+                                $tmp1$$FloatRegister, $tmp2$$FloatRegister);
+  %}
+  ins_pipe(pipe_slow);
+%}
+
+instruct reduce_mulL(iRegLNoSp dst, iRegL isrc, vReg vsrc) %{
+  predicate(Matcher::vector_length_in_bytes(n->in(2)) == 16);
+  match(Set dst (MulReductionVL isrc vsrc));
+  effect(TEMP_DEF dst);
+  format %{ "reduce_mulL $dst, $isrc, $vsrc\t# 2L" %}
+  ins_encode %{
+    __ neon_reduce_mul_integral($dst$$Register, T_LONG, $isrc$$Register,
+                                $vsrc$$FloatRegister, 16, fnoreg, fnoreg);
+  %}
+  ins_pipe(pipe_slow);
+%}
+
+instruct reduce_mulF(vRegF dst, vRegF fsrc, vReg vsrc, vReg tmp) %{
+  predicate(Matcher::vector_length_in_bytes(n->in(2)) <= 16);
+  match(Set dst (MulReductionVF fsrc vsrc));
+  effect(TEMP_DEF dst, TEMP tmp);
+  format %{ "reduce_mulF $dst, $fsrc, $vsrc\t# 2F/4F. KILL $tmp" %}
+  ins_encode %{
+    uint length_in_bytes = Matcher::vector_length_in_bytes(this, $vsrc);
+    __ neon_reduce_mul_fp($dst$$FloatRegister, T_FLOAT, $fsrc$$FloatRegister,
+                          $vsrc$$FloatRegister, length_in_bytes, $tmp$$FloatRegister);
+  %}
+  ins_pipe(pipe_slow);
+%}
+
+instruct reduce_mulD(vRegD dst, vRegD dsrc, vReg vsrc, vReg tmp) %{
+  predicate(Matcher::vector_length_in_bytes(n->in(2)) == 16);
+  match(Set dst (MulReductionVD dsrc vsrc));
+  effect(TEMP_DEF dst, TEMP tmp);
+  format %{ "reduce_mulD $dst, $dsrc, $vsrc\t# 2D. KILL $tmp" %}
+  ins_encode %{
+    __ neon_reduce_mul_fp($dst$$FloatRegister, T_DOUBLE, $dsrc$$FloatRegister,
+                          $vsrc$$FloatRegister, 16, $tmp$$FloatRegister);
+  %}
+  ins_pipe(pipe_slow);
+%}
+
+dnl
+dnl REDUCE_BITWISE_OP_NEON($1,        $2       $3    $4     )
+dnl REDUCE_BITWISE_OP_NEON(insn_name, is_long, type, op_name)
+define(`REDUCE_BITWIESE_OP_NEON', `
```

Comment:
`REDUCE_BITWIESE_OP_NEON` doesn't look right,

Reply:
Thanks for pointing this out! Updated here and other similar macros.